timemachine/src/time_travel.rs

```diff
@@ -0,0 +1,742 @@
+      while let Some(incoming) = incoming.next().await {
+        let (key, range) = incoming?;
+        let tx = self.db.transaction()?;
+
+        // Check again in the transaction to make sure we don't overwrite new data
+        let pulled_now = tx.query_row(
+          "select pulled from tt_initial_snapshot_ranges where format_version = ? and monoseq = ? and seq = ?",
+          rusqlite::params![&key.format_version, &key.monoseq, &key.seq],
+          |row| Ok(row.get::<_, u64>(0)? != 0),
+        )?;
+        if pulled_now {
+          continue;
+        }
+
+        for (data_entry, metadata_entry) in
+          range.data_list.iter().zip(range.metadata_list.iter())
+        {
+          assert_eq!(data_entry.key[0], b'd');
+          assert_eq!(metadata_entry.key[0], b'm');
+          assert_eq!(data_entry.key[1..], metadata_entry.key[1..]);
+
+          let key = &data_entry.key[1..];
+          let value = &data_entry.value[..];
+          let metadata = KeyMetadata::decode(&metadata_entry.value[..])
+            .ok_or_else(|| {
+              anyhow::anyhow!("invalid metadata for key {}", hex::encode(key))
+            })?;
+          tx.execute(
+          "insert into kv_snapshot (k, v, v_encoding, version, seq) values (?, ?, ?, ?, ?)
+          on conflict(k) do update set v = excluded.v, v_encoding = excluded.v_encoding, version = excluded.version, seq = excluded.seq",
+          rusqlite::params![
+            key,
+            value,
+            metadata.value_encoding,
+            u64::from_be_bytes(<[u8; 8]>::try_from(&metadata.versionstamp[0..8]).unwrap()),
+            u16::from_be_bytes(<[u8; 2]>::try_from(&metadata.versionstamp[8..10]).unwrap()),
+          ],
+        )?;
+        }
+        tx.execute(
+          "update tt_initial_snapshot_ranges set pulled = 1 where format_version = ? and monoseq = ? and seq = ?",
+          rusqlite::params![&key.format_version, &key.monoseq, &key.seq],
+        )?;
+        tx.commit()?;
+      }
+    }
+
+    // Step 3: Fetch REDO logs until DV is reached and we are up to date as measured
+    // in real time
+    let start_ts = SystemTime::now()
+      .duration_since(SystemTime::UNIX_EPOCH)
+      .unwrap()
+      .as_millis() as u64;
+    let last_key = self
+      .fetch_redo_outside_tx(source, |x| {
+        x.key.first_versionstamp12 > dv && x.last_modified_ms > start_ts
+      })
+      .await?;
+
+    // Step 4: Apply REDO to bring kv_snapshot to DV
+    {
+      let tx = self.db.transaction()?;
+      if get_config(&tx, TT_CONFIG_KEY_INITIAL_SNAPSHOT_COMPLETED)?.as_deref()
+        == Some("1")
+      {
+        return Ok(());
+      }
+
+      set_config(
+        &tx,
+        TT_CONFIG_KEY_CURRENT_SNAPSHOT_VERSIONSTAMP12,
+        &hex::encode([0u8; 12]),
+      )?;
+      if last_key.is_some() {
+        replay_redo_or_undo_in_tx(&tx, dv, false)?;
+      }
+      set_config(&tx, TT_CONFIG_KEY_INITIAL_SNAPSHOT_COMPLETED, "1")?;
+      tx.commit()?;
+    }
```

Comment:
It wasn't clear to me after first reading this that step 3 syncs the logs into sqlite, and then this line seeks the view to the correct log entry. Maybe a couple more comments would be useful.